### PR TITLE
content: add feature page for provider physical network

### PR DESCRIPTION
**Pull request by phoracek**

### CODE REVIEW

**Comment by phoracek** [comment]
@mkalfon please review

**Comment by phoracek** [comment]
@mmirecki please review

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+feature_name: Provider Physical Network
+feature_modules: engine,vdsm,ovn-provider
+feature_status: In Development
+---
+
+# Provider Physical Network
+
+## Summary
+
+This feature extends [ovirt-ovn-provider](https://www.ovirt.org/develop/release-management/features/network/ovirt-ovn-provider/) supporting physical network access. With this feature, user will be able connect external OVN network to a physical (oVirt native) network.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: maybe:
supporting physical network access -> adding support for physical network access.

With this feature, user will be able connect external OVN network to a physical (oVirt native) network.
->
It allows to to connect external OVN networks to a physical (oVirt native) network.

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+Progress of the feature is tracked on [ovn-localnet Trello](https://trello.com/b/lxIBEn0A/ovn-localnet). Related patches could be found on [gerrit topic:localnet](https://gerrit.ovirt.org/#/q/topic:localnet).
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: Progress of the feature -> Feature progress

could be -> can be

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+### Owner
+
+- Feature Owner: Petr Horáček (phoracek)
+- E-mail: <phoracek@redhat.com>
+
+### Benefit to oVirt
+
+With this feature existing OVS VM networking could be replaced with OVN, providing traditional oVirt networking. That would also allow us to drop a part of VM live migration hook.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: With this feature existing OVS VM networking could be replaced with OVN.
->
Existing OVS VM networking could be replaced with OVN.

providing traditional oVirt networking.
-> replacing the legacy, bridge based oVirt networking.
(am I getting the intended meaning right?)

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+Physical network access will be required once NAT and routing are introduced OVN provider.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: introduced tp the OVN provider.

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+## Usage
+
+### Set cluster network switch type to OVS
+
+This feature requires cluster switch type set to OVS. You can enable OVS networking with following steps.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: with following -> using the following

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+1) **Enable OVS hook for live migration on all vdsm hosts.** Add `migration_ovs_hook_enabled = true` to `[vars]` section in `/etc/vdsm/vdsm.conf`. And restart vdsm with `systemctl restart vdsmd supervdsmd`.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: And restart vdsm with
->
Restart vdsm using:

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+1) **Set Cluster switch type to OVS.** Open `Edit` on selected cluster and set `Switch Type` to `OVS (experimental)`.
+
+1) **Set OVS networking on all vdsm hosts.** Go over each Host, put it to `Maintenance`, `Sync All Networks` and `Activate`.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: Go over each Host, put it to
->
For each Host, enable

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+### Create an external network on top of a physical network
+
+There are several ways how to use this feature.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: remove 'how'

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+- **Select physical network via custom values.** When creating a new external network, fill physical network name to `Physical Network`. If the physical network has VLAN tagging enabled, also check `Enable VLAN tagging` and copy physical network's tag there. Please note, that physical network name references to VDSM network name, which is not necessarily the same as oVirt network name (in case it has more than 15 characters or contains special characters).
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: fill physical network name to `Physical Network`
-> set  `Physical Network` to the physical network name.

also check - > check

and copy physical -> and set the physical 

references to VDSM network name-> references a VDSM network name

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+![add an external network connected to a physical network Engine dialog](/images/features/network/provider-physical-network_new-network-dialog-engine.png)
+
+- **Select physical network from data center networks.** When creating a new external network, in `Physical Network` section, select `Data Center Network` and pick desired oVirt network from drop down list. Engine then uses VDSM network name and VLAN ID from the selected network. This can also be done via REST API using `<external_provider_physical_network id="123"/>`, see [ovirt-engine-api-model documentation](http://ovirt.github.io/ovirt-engine-api-model/4.2/#types/network).
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: in `Physical Network` section, select `Data Center Network` 
->  select `Data Center Network` in the `Physical Network` section,

pick desired -> pick the desired

from drop down list -> from the drop down list

uses VDSM network name -> uses the VDSM network name

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 9 above)
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: "via custom values"
This sounds as if there were multiple custom values, like for DNS's.
maybe change this to:
-> via a custom physical network name
or
-> via a custom value

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+- **Create external network implicitly with physical network.** To make this feature more exposed, we implicitly create a matching external OVN network for each new OVS network marked as `VM Network`. However, there is no connection between them afterwards, they behave as separate units.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: "However, there is no connection between them afterwards, they behave as separate units."
So what is the sense of creating the matching OVN network? Do we not create it to allow using the OVS network as a physical network?
If there is no connection why do we include it in the list of how this feature can be used?

Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: OVS network? I assume we mean an ovirt OVS network (OVS bridge).
Could we make this more clear?

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+- **Attach external network via custom values in ManageIQ**. This option is similar to the first one. In ManageIQ a user can set network type `flat` and specify physical network name (VDSM network name), in case physical network is on VLAN, set type to `vlan` and specify VLAN ID as well.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: name), in - > name). In

is on VLAN -> is on a VLAN

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+![add an external network connected to a physical network ManageIQ dialog](/images/features/network/provider-physical-network_new-network-dialog-miq.png)
+
+### Attach VM to the external network
+
+Finally add a new NIC to a VM and select a profile of an external network. L2 connectivity should be provided to the VM. In case there is DHCP server running, VM should obtain an IP.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: is DHCP server -> is a DHCP server

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+## Caveat
+
+User needs to make sure that physical network used by external network is available on the host used by VM. There is currently now way for Engine how to enforce it. Easiest way how to accomplish this is to set physical network as required for the cluster.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: User -> The user
 that physical network ->  that the physical network
 external network  ->  the external network 
VM -> the VM
drop "how" -> Engine to enforce
Easiest way how to accomplish -> The easiest way to accomplish

Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: "to set physical network as required for the cluster."
This suggests we set a physical network at cluster level. Is this what we want to say?

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+## Implementation
+
+### OpenStack Neutron API
+
+OpenStack Neutron API used by Engine to control external providers already provides means to specify physical network – attributes `provider:network_type`, `provider:physical_network` and `provider:segmentation_id`. Network types we use are `flat` and `vlan`. Physical network is matched by VDSM network name. Segmentation ID is used with VLAN networks and matches VLAN ID defined for selected network.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: specify physical network -> specify a physical network
Network types -> The network types
 VDSM network  ->  the VDSM network 
VLAN ID  -> the VLAN ID 
selected network - > the selected network

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+### oVirt Provider OVN
+
+Provider reads attributes described above. If physical network is defined, provider implicitly creates `localnet` port on given network and attaches it to passed physical network. In case segmentation ID was passed, it uses it as a VLAN tag on the port.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: The provider reads the attributes described above. If the physical network is defined, the provider implicitly creates a `localnet` port on the given network and attaches it to the received physical network. In case a segmentation ID was passed, it uses it as the VLAN tag on the port.

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+### VDSM
+
+Localnet port is attached to an abstract network name. This name is mapped to an OVS bridge name on each host via OVS DB attribute `external-ids:ovn-bridge-mappings`. This attribute is configured after each `setupNetworks` command and also during upgrade (after reboot or upgrade of vdsm package).
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: drop "also"
upgrade -> upgrades

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+### Engine
+
+`ProviderNetwork` object in Engine is extended with a link to its physical network (`Network`). In database, `provider_physical_network_id` is added as a new column to `network` table.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: The 'Provider Network'
the database

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+
+Physical network is used to find VDSM network name and VLAN ID, these values are then passed to network provider.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: Physical network is used to find the VDSM network name and the VLAN ID. These values are then passed to the network provider.

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+
+Engine also covers validation: physical network is on the same Data Center as the external network, custom values (physical network and VLAN ID) are not specified when provider physical network is set.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: Engine also covers the validation: 
- physical network must be on the same Data Center as the external network
- the custom values (physical network and VLAN ID) must not be specified a when provider physical network is set.

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+### Engine UI
+
+Without this feature, there is already a way how to set physical network name and VLAN ID for an external network from `New Network` dialog. Physical network has its separate field, VLAN is obtained from shared `Network Attributes` section. However, this way requires user to know VDSM name of the network and also manually copy VLAN ID of the network.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: Even without this feature, it is already possible to set the physical network name and VLAN ID for an external network from `New Network` dialog. Physical network has a separate field, and the VLAN is obtained from the shared `Network Attributes` section. However, this way requires the user to know the VDSM name of the network and manually copy the VLAN ID of the network

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+
+For better user experience, this feature exposes an option to select desired provider physical network from Data Center networks to UI.
+
+![add an external network connected to a physical network dialog](/images/features/network/provider-physical-network_new-network-dialog.png)
+
+To expose this feature, external OVN network is created implicitly when user adds a new VM network on OVS cluster. When selecting a network profile for a VM NIC, we list only external networks, not native oVirt networks that triggered their creation.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: To expose this feature, an external OVN network is created implicitly when the user adds a new VM network on an OVS cluster. When selecting a network profile for a VM NIC, we list only external networks, not native oVirt networks that triggered their creation.

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+
+### REST API
+
+`Network` object in REST API is now extended with two new links:
+
+`<external_provider id="123"/>` points to a `OpenStackNetworkProvider`. If this link is specified, new network is defined on the selected provider. This mimics Engine UI `New Network` dialog with `Create on external provider` checked.
```
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: a -> an

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 9 above)
Comment (mmirecki) [inline source/develop/release-management/features/network/provider-physical-network.md]: Will the custom values not be replaced by what is shown on: provider-physical-network_new-network-dialog.png

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 3 above)
Comment (phoracek) [reply source/develop/release-management/features/network/provider-physical-network.md]: I'm not sure, i'm trying to say that current OVS switch type networking will use OVN for VM networks rather than OVS bridging.

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 11 above)
Comment (phoracek) [reply source/develop/release-management/features/network/provider-physical-network.md]: we do, i mean there is no connection between their respective engine objects

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 14 above)
Comment (phoracek) [reply source/develop/release-management/features/network/provider-physical-network.md]: physical provider network then?

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 9 above)
Comment (phoracek) [reply source/develop/release-management/features/network/provider-physical-network.md]: I don't understand this comment

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+Physical network access will be required once NAT and routing are introduced to the OVN provider.
+
+## Usage
+
+### Set cluster network switch type to OVS
+
+This feature requires cluster switch type set to OVS. You can enable OVS networking using the following steps.
+
+1) **Enable OVS hook for live migration on all vdsm hosts.** Add `migration_ovs_hook_enabled = true` to `[vars]` section in `/etc/vdsm/vdsm.conf`. Restart vdsm with `systemctl restart vdsmd supervdsmd`.
```
Comment (dankenigsberg) [inline source/develop/release-management/features/network/provider-physical-network.md]: why is this required? I think that with OVN-only networks it is not needed at all.

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+Feature progress is tracked on [ovn-localnet Trello](https://trello.com/b/lxIBEn0A/ovn-localnet). Related patches can be found on [gerrit topic:localnet](https://gerrit.ovirt.org/#/q/topic:localnet).
+
+### Owner
+
+- Feature Owner: Petr Horáček (phoracek)
+- E-mail: <phoracek@redhat.com>
+
+### Benefit to oVirt
+
+Existing OVS VM networking could be replaced with OVN. That would also allow us to drop a part of VM live migration hook.
```
Comment (dankenigsberg) [inline source/develop/release-management/features/network/provider-physical-network.md]: which part is still required? you've already killed the display network part.

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 24 above)
Comment (dankenigsberg) [inline source/develop/release-management/features/network/provider-physical-network.md]: do we really need this?

File: source/develop/release-management/features/network/provider-physical-network.md [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+### Engine UI
+
+Even without this feature, it is already possible to set the physical network name and VLAN ID for an external network from `New Network` dialog. Physical network has a separate field, and the VLAN is obtained from the shared `Network Attributes` section. However, this way requires the user to know the VDSM name of the network and manually copy the VLAN ID of the network
+
+For better user experience, this feature exposes an option to select desired provider physical network from Data Center networks to UI.
+
+![add an external network connected to a physical network dialog](/images/features/network/provider-physical-network_new-network-dialog.png)
+
+To expose this feature, an external OVN network is created implicitly when the user adds a new VM network on an OVS cluster. When selecting a network profile for a VM NIC, we list only external networks, not native oVirt networks that triggered their creation.
+
```
Comment (dankenigsberg) [inline source/develop/release-management/features/network/provider-physical-network.md]: please add a note about two future additions:
* https://trello.com/c/Uj9xobnx/5-disallow-direct-connection-of-vm-to-a-vm-network-in-ovs-cluster
* https://trello.com/c/hVIVxjB7/24-when-defining-a-physical-network-auto-define-an-external-network-based-on-it

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 25 above)
Comment (phoracek) [reply source/develop/release-management/features/network/provider-physical-network.md]: the VM network part. we need to change OVS bridge name on migration. but since it does not work now and we will drop it (?) we can just drop this note

File: source/develop/release-management/features/network/provider-physical-network.md (hunk 24 above)
Comment (phoracek) [reply source/develop/release-management/features/network/provider-physical-network.md]: For physical networks it is needed. We need an OVS bridge to connect to.

Comment (phoracek) [reply source/develop/release-management/features/network/provider-physical-network.md]: We don't and it does not work anyways.
